openedx/core/djangoapps/programs/tasks/v1/tasks.py

```diff
@@ -331,3 +350,173 @@ def award_course_certificate(self, username, course_run_key):
+        username: The username of the student
+        program_uuid: uuid of the program
+
+    Returns:
+        None
+
+    """
+    client.credentials.post({
+        'username': username,
+        'status': 'revoked',
+        'credential': {
+            'type': PROGRAM_CERTIFICATE,
+            'program_uuid': program_uuid
+        }
+    })
+
+
+@task(bind=True, ignore_result=True, routing_key=PROGRAM_CERTIFICATES_ROUTING_KEY)
+def revoke_program_certificates(self, username, course_key):
+    """
+    This task is designed to be called whenever a student's course certificate is
+    revoked.
+
+    It will consult with a variety of APIs to determine whether or not the
+    specified user's certificate should be revoked in one or more programs, and
+    use the credentials service to revoke the said certificates if so.
+
+    Args:
+        username (str): The username of the student
+        course_key (str|CourseKey): The course identifier
+
+    Returns:
+        None
+
+    """
+    countdown = 2 ** self.request.retries
+    # If the credentials config model is disabled for this
+    # feature, it may indicate a condition where processing of such tasks
+    # has been temporarily disabled.  Since this is a recoverable situation,
+    # mark this task for retry instead of failing it altogether.
+
+    if not CredentialsApiConfig.current().is_learner_issuance_enabled:
+        LOGGER.warning(
+            'Task revoke_program_certificates cannot be executed when credentials issuance is disabled in API config',
+        )
+        raise self.retry(countdown=countdown, max_retries=MAX_RETRIES)
+
+    try:
+        try:
+            student = User.objects.get(username=username)
+        except User.DoesNotExist:
+            LOGGER.exception(u'Task revoke_program_certificates was called with invalid username %s', username)
+            # Don't retry for this case - just conclude the task.
+            return
+
+        inverted_programs = get_inverted_programs(student)
+        course_specific_programs = inverted_programs.get(str(course_key))
+        if not course_specific_programs:
+            # No reason to continue beyond this point
+            LOGGER.info(
+                u'Task revoke_program_certificates was called for user %s and course %s with no engaged programs',
+                username,
+                course_key
+            )
+            return
+
+        # Determine which program certificates the user has already been awarded, if any.
+        program_uuids_to_revoke = get_revokable_program_uuids(course_specific_programs, student)
+    except Exception as exc:
+        LOGGER.exception(
+            u'Failed to determine program certificates to be revoked for user %s with course %s',
+            username,
+            course_key
+        )
+        raise self.retry(exc=exc, countdown=countdown, max_retries=MAX_RETRIES)
+
+    if program_uuids_to_revoke:
+        try:
```

Comment:
it would be nice if revoking part has its separate function call, here it is creating celery task too messy! Just a suggestion!

Reply:
A good suggestion but I tried to use same logic as used in other two tasks here, and already moved the revoking call in a separate function. I think if we are going to do this refactor  then we should do it for other tasks as well but it will go out of scope of this ticket.